### PR TITLE
Fix subscription processing with mandates

### LINKED ISSUE #3230: Variable subscriptions unaccounted for when creating mandates
**Describe the bug**
Indian credit cards require autopay mandate to be setup. The plugin is currently setting up the mandate amount same as subscription price. If the user upgrades to a higher plan with more price, then the renewal fails. 

This happens because the earlier mandate is not valid anymore.

**To Reproduce**
1. Create variable subscription with 2 variations with prices ₹1,000/mon and ₹2,000/moon.
2. First buy ₹1,000/mon subscription on say 1st of the month. This will setup autopay mandate of ₹1,000.
3. Now upgrade to the ₹2,000/mon subscription in the middle of the month. Switch will happen successfully but the autopay mandate is still ₹1,000.
4. On the renewal day the payment will fail as mandate amount is less than ₹2,000.

**Expected behavior**
New mandate should be set during switching, that way the switching will happen without failing.

**Environment (please complete the following information):**
 - WordPress Version: 6.5.5
 - WooCommerce Version: 9.0.2
 - Stripe Plugin Version: 8.4.0
 - WooCommerce Subscriptions Plugin Version: 6.4.0

### PR DESCRIPTION
**Pull request by wjrosa**
Fixes #3230

## Changes proposed in this Pull Request:

This PR fixes the processing of subscriptions when using mandates (i.e., when using Indian credit cards). 

### Renewal

- Instructions based [on the testing steps of the original issue](https://github.com/woocommerce/woocommerce-gateway-stripe/issues/3230)
- Checkout to this branch on your local environment (`fix/subscriptions-switch-with-mandates`)
- Set your store and Stripe account currency to `Indian Rupee`
- Install and enable the subscriptions extension, and enable switching on `wp-admin/admin.php?page=wc-settings&tab=subscriptions`
- Create a variable subscription with two variations with prices ₹1,000/month and ₹2,000/month.
- First, buy ₹1,000/month subscription on, say, the 1st of the month. This will set up an autopay mandate of ₹1,000.
- Now upgrade to the ₹2,000/month subscription in the middle of the month *. The switch will happen successfully
- On the renewal day*, the payment should succeed
- On the `develop` branch it should fail

\* You can force the renewal on the subscription details page (wp-admin) by selecting "Process renewal" in the dropdown box. @james-allan gave some tips on how to easily test this [here](https://github.com/woocommerce/woocommerce-gateway-stripe/pull/3323#issuecomment-2275153628).

### Subscription edit

- With your store currency set to `Indian Rupee`, purchase a subscription
- Confirm that you cannot edit this subscription on wp-admin (it is not possible to replace it with another subscription product)

---

-   [ ] Covered with tests (or have a good reason not to test in description ☝️)
-   [x] Added changelog entry **in both** `changelog.txt` and `readme.txt` (or does not apply)
-   [x] Tested on mobile (or does not apply)

**Post merge**

-   [ ] Added testing instructions to the [Release Testing Instructions wiki page](https://github.com/woocommerce/woocommerce-gateway-stripe/wiki/Release-Testing-Instructions) (or does not apply)
-   [ ] Added the [needs docs label](https://github.com/woocommerce/woocommerce-gateway-stripe/labels?q=docs) (or does not apply)
-   [ ] Included this PR in the Release Thread scope (or does not apply)

### CODE REVIEW

File: includes/compat/trait-wc-stripe-subscriptions.php [hunk 1]
```diff
@@ -674,19 +682,40 @@ private function create_mandate_options_for_order( $order, $subscriptions ) {
...
 
-		// If there are no subscriptions we just return since mandates aren't required.
-		if ( 0 === count( $subscriptions ) ) {
-			return [];
-		}
+		// If this is a switch order we set the mandate options based on the new subscription.
+		$cart_contain_switches = WC_Subscriptions_Switcher::cart_contains_switches();
+		if ( $cart_contain_switches ) {
+			foreach ( WC()->cart->cart_contents as $cart_item ) {
+				$sub_amount += (int) WC_Subscriptions_Product::get_price( $cart_item['data'] );
```
Comment (Mayisha) [inline includes/compat/trait-wc-stripe-subscriptions.php]: We may need to consider different decimal currencies like [this](https://github.com/woocommerce/woocommerce-gateway-stripe/blob/develop/includes/compat/trait-wc-stripe-subscriptions.php#L688-L690)?

Comment (wjrosa) [reply includes/compat/trait-wc-stripe-subscriptions.php]: Good catch, Mayisha! I fixed it here and also [here](https://github.com/woocommerce/woocommerce-gateway-stripe/pull/3359/files#diff-516e329017ff60072191226cd6dfc80b4b9ca8ccccacd6c151c1598cf4179434R712) (which ignored the currency).

File: includes/compat/trait-wc-stripe-subscriptions.php [hunk 2]
```diff
@@ -87,6 +87,9 @@ public function maybe_init_subscriptions() {
 		*/
 		add_action( 'template_redirect', [ $this, 'remove_order_pay_var' ], 99 );
 		add_action( 'template_redirect', [ $this, 'restore_order_pay_var' ], 101 );
+
+		// Disable editing for Indian subscriptions with mandates. Those need to be recreated as mandates does not support upgrades (due fixed amounts).
+		add_filter( 'wc_order_is_editable', [ $this, 'disable_subscription_edit_for_india' ], 10, 2 );
```
Comment (Mayisha) [inline includes/compat/trait-wc-stripe-subscriptions.php:92]: This is a very good idea. 👍

**Review by Mayisha — APPROVED** [review]
Nice work This PR perfectly fixed the issue of switching variable subscriptions. @wjrosa 🎉 

I have noticed a few problems while testing.
- No mandate is created when I use a saved card.
- Renewal fails when the shipping charge of the subscription order and the recurring order are different. Suppose the subscription price is 1000, the initial shipping charge is free and the recurring shipping charge is 20. Then it fails with the following error-

`
payment_intent_mandate_invalid
This mandate can only be used for a payment of 12000. You attempted a payment of 10000.
`
- The text is wrong when the subscription is not editable.
Screenshot 2024-08-28 at 10 14 44 PM

Approving this PR as I can reproduce the above issues in the `develop` branch as well. I will create separate issues to fix them.

**Comment by Mayisha** [comment]
Issues created

1. https://github.com/woocommerce/woocommerce-gateway-stripe/issues/3393
2. https://github.com/woocommerce/woocommerce-gateway-stripe/issues/3394
3. https://github.com/Automattic/woocommerce-subscriptions-core/issues/665

**Comment by wjrosa** [comment]
Thank you, @Mayisha ! I will grab some of these

**Pull request merged**